### DISCUSSION TITLE
Sharing cache on Windows workers get systematically low hits

### DISCUSSION THREAD

**Discussion by Esysc** [ccache/ccache#1084] [Support]
We use widely `ccache` in combination with `cache-s3` to reduce compilation time, to share the cache and it works fine. The platforms where is currently in production are Linux and MacOS. We build also on Windows and until now we are using `clcache`. As this project is archived and unmaintained we are  trying to migrate to `ccache` that is considerately increasing Windows support. Despite several configuration changes and tries, I didn't succeed yet to get high rate of cache hits. Obviously, if I run  another build on the same host, the cache gets high hits and the compilation goes very fast.

## Context

- `ccache` version 4.6.1 (latest at the moment of writing)
- CI system [Concourse](https://github.com/concourse/concourse)  
- the build is done out of source in `...\containers\<random-id>\tmp\build\734c03a2\build-windows`
- the source is in `...\containers\<random-id>\tmp\build\734c03a2\repo.git`
- the path `...\containers\<random-id>\tmp\build\734c03a2 is a symlink to ...\volumes\live\<volume-uuid>\volume`
- the path `...\containers\<random-id>\tmp\build\734c03a2\repo.git` is a symlink to ...\volumes\live\<another-volume-uuid>\volume`
- the `...` part above are common to all builds on all CI nodes
- `<random-id>`, `<volume-uuid>` and `<another-volume-uuid>` are the variable part of paths across builds.

Builds run on Concourse CI Windows workers.
The cache is saved to and restored from a S3 bucket (https://github.com/fpco/cache-s3)

### Try 1
In my first attempt, as documented, I set `CCACHE_BASEDIR` to the build root directory path and `CCACHE_DIR` inside it.
```
set CCACHE_BASEDIR=C:\Users\<ci-user>\worker\containers\<random-id>\tmp\build\734c03a2
set CCACHE_DIR=C:\Users\<ci-user>\worker\containers\<random-id>\tmp\build\734c03a2\ccache
```
With this configuration the absolute paths should be rewritten to relative paths as far as understand, permitting successive builds to consume the cache, but it's not working: in the debug logs I see the some includes paths are not normalized. 
Below, a snippet of the commandline.
```
 Command line: C:\ccache\ccache.exe C:\PROGRA2\MICROS1\2019\PROFES1\VC\Tools\MSVC\14291.301\bin\Hostx64\x64\cl.exe /nologo .... -IC:\Users\<ci-user>\worker\containers\<random-id>\tmp\build\734c03a2\repo.git\src\pix4d_calibration_processing\include ....
```
Small extract of input text log, path is not normalized:
```
...
Note: including file:   ../repo.git/src/api/calibrationdatamodel/include\calibrationdatamodel/uncertainties.h

Note: including file: C:\Users\<ci-user>\worker\containers\<random-id>\tmp\build\734c03a2\repo.git\src\api\calibrationapi\src\alignmentTransformationHelper.h
...
```
It seems that include wrapped ion double quotes are **not normalized**
```
#include <calibrationapi/alignmentTransformation.h>

#include "alignmentTransformationHelper.h"
```

And below the debug log
```
[2022-05-24T15:12:13.033713 4716 ] Hostname: EC2AMAZ-F2RHDDV
[2022-05-24T15:12:13.033713 4716 ] Working directory: C:/Users/<ci-user>/worker/containers/<random-id>/tmp/build/734c03a2/build
[2022-05-24T15:12:13.033713 4716 ] Compiler type: msvc
[2022-05-24T15:12:13.033713 4716 ] Source file: ../repo.git/src/api/calibrationapi/src/alignmentTransformation.cpp
[2022-05-24T15:12:13.033713 4716 ] Object file: src\api\calibrationapi\CMakeFiles\calibrationapi.dir\src\alignmentTransformation.cpp.obj
[2022-05-24T15:12:13.049338 4716 ] Hashing object file directory C:/Users/<ci-user>/worker/containers/00003of24si/tmp/build/734c03a2/build
[2022-05-24T15:12:13.049338 4716 ] Trying direct lookup
[2022-05-24T15:12:13.128059 4716 ] No 8843bf35skpnmg4lpp5b6lb2s409j7ja6 in primary storage
[2022-05-24T15:12:13.128059 4716 ] Running preprocessor
[2022-05-24T15:12:15.577627 4716 ] Got result key from preprocessor
[2022-05-24T15:12:15.577627 4716 ] No ba12nm1mst25q41mbjgjrgsa5usa8fhfu in primary storage
[2022-05-24T15:12:15.577627 4716 ] Running real compiler
[2022-05-24T15:12:23.124578 4716 ] Using default compression level 1
[2022-05-24T15:12:23.124578 4716 ] Storing embedded entry #0 <stdout> (126337 bytes)
[2022-05-24T15:12:23.124578 4716 ] Storing embedded entry #1 .o (336356 bytes) from src\api\calibrationapi\CMakeFiles\calibrationapi.dir\src\alignmentTransformation.cpp.obj
[2022-05-24T15:12:23.124578 4716 ] Stored ba12nm1mst25q41mbjgjrgsa5usa8fhfu in primary storage (C:\Users\<ci-user>\worker\containers\<random-id>\tmp\build\734c03a2\ccache/b/a/12nm1mst25q41mbjgjrgsa5usa8fhfuR)
[2022-05-24T15:12:23.124578 4716 ] Adding result key to C:\Users\<ci-user>\worker\containers\<random-id>\tmp\build\734c03a2\ccache/8/8/43bf35skpnmg4lpp5b6lb2s409j7ja6M
[2022-05-24T15:12:23.140229 4716 ] Using default compression level 1
[2022-05-24T15:12:23.140229 4716 ] Stored 8843bf35skpnmg4lpp5b6lb2s409j7ja6 in primary storage (C:\Users\concourse\worker\containers\<random-id>\tmp\build\734c03a2\ccache/8/8/43bf35skpnmg4lpp5b6lb2s409j7ja6M)
[2022-05-24T15:12:23.140229 4716 ] Result: cache_miss
[2022-05-24T15:12:23.140229 4716 ] Result: direct_cache_miss
[2022-05-24T15:12:23.140229 4716 ] Result: preprocessed_cache_miss
[2022-05-24T15:12:23.140229 4716 ] Result: primary_storage_miss
[2022-05-24T15:12:23.140229 4716 ] Result: primary_storage_miss
[2022-05-24T15:12:23.140229 4716 ] Acquired lock C:\Users\<ci-user>\worker\containers\<random-id>\tmp\build\734c03a2\ccache/8/stats.lock
[2022-05-24T15:12:23.140229 4716 ] Releasing lock C:\Users\<ci-user>\worker\containers\<random-id>\tmp\build\734c03a2\ccache/8/stats.lock
[2022-05-24T15:12:23.155854 4716 ] Acquired lock C:\Users\<ci-user>\worker\containers\<random-id>\tmp\build\734c03a2\ccache/b/stats.lock
[2022-05-24T15:12:23.155854 4716 ] Releasing lock C:\Users\<ci-user>\worker\containers\<random-id>\tmp\build\734c03a2\ccache/b/stats.lock
```

### Try 2
In my second attempt, after reading this chapter  https://ccache.dev/manual/4.6.1.html#_sharing_a_cache, I modified CCACHE_DIR to a fixed common path for all builds.
```
set CCACHE_DIR=C:\temp\734c03a2
```
It didn't work neither, I got the same low hits statistics.

### Try 3
I did a third attempt setting `CCACHE_BASEDIR` and `CCACHE_DIR` to the target value instead of symlink guessing something wrong on how these pointer are managed on Windows.
```
set CCACHE_BASEDIR=C:\Users\<ci-user>\worker\volumes\live\<volume-uuid>\volume
set CCACHE_DIR=C:\Users\<ci-user>\worker\volumes\live\<volume-uuid>\volume\ccache
```
Same discouraging results. 

### Try 4
I did a fourth attempt: in my build script I get the real path of the target, then I create myself a symlink that is constant for every build on different hosts:
```
set CCACHE_BASEDIR=C:\734c03a2       => soft link to  C:\Users\<ci-user>\worker\volumes\live\<volume-uuid>\volume
set CCACHE_DIR=C:\734c03a2\ccache
```

In all these four cases, the statistics are quite the same:

```
"Compiler cache statistics"
Summary:
  Hits:              29 /  881 (3.29 %)
    Direct:           0 /  881 (0.00 %)
    Preprocessed:    29 /  881 (3.29 %)
  Misses:           852
    Direct:         881
    Preprocessed:   852
Primary storage:
  Hits:              29 / 1762 (1.65 %)
  Misses:          1733
  Cache size (GB): 0.08 / 5.00 (1.61 %)

Use the -v/--verbose option for more details.
```

### Some extras
While executing the 4 tries above, I played with some additional environment settings:
```
set CCACHE_COMPILERCHECK=content
set CCACHE_SLOPPINESS=time_macros
```
I did without, with both, with just one, but it seems it's the `PATH` the reason why cache has low hits.

## Conclusion

Is `ccache` completely supported on Windows platforms for sharing having these path drifts between builds? 
Is it a `ccache` bug the fact that some paths when they are wrapped in double quotes are not normalized?

I'm asking if there are  more experienced users that already are doing what I'm trying and could give me some hints.

**Comment by jrosdahl** [comment]
> It seems that include wrapped ion double quotes are **not normalized**

To get more information on this, please enable [CCACHE_DEBUG](https://ccache.dev/manual/4.6.1.html#_cache_debugging) and check the corresponding `.ccache-input-text` file to see if there is any difference in the preprocessor output for the two cases.

> [2022-05-24T15:12:13.049338 4716 ] Hashing object file directory C:/Users/<ci-user>/worker/containers/00003of24si/tmp/build/734c03a2/build

This means that the volatile directory name is included in the hash, so there won't be any hits when compiled in another directory. Here is why this is done:

https://github.com/ccache/ccache/blob/b36a81893c6f2cd20087815df2ddeb56dd547813/src/ccache.cpp#L1372-L1374

You can opt out of this by setting `CCACHE_NOHASHDIR`.

> I did a third attempt setting `CCACHE_BASEDIR` and `CCACHE_DIR` to the target value instead of symlink guessing something wrong on how these pointer are managed on Windows.

`CCACHE_DIR` should typically be in a place with a known non-volatile path, since how can it otherwise be reused between builds? Or is this handled by `cache-s3` (not familiar with this)? That said, the cache location does not affect how the input hash is calculated.

`CCACHE_BASEDIR` can be any parent directory of the build tree. The value itself is not part of the hash, and you don't have to fiddle with symlinks (if that's what you mean by "soft link").

> Is `ccache` completely supported on Windows platforms for sharing having these path drifts between builds?

The Windows support is driven by the ccache community, and nobody has (as far as I know) evaluated how much of the feature set works on Windows.

> Is it a `ccache` bug the fact that some paths when they are wrapped in double quotes are not normalized?

Probably that or a missing feature. :slightly_smiling_face:

**Comment by Esysc** [comment]
Thank you @jrosdahl for your response, I actually set `CCACHE_NOHASHDIR` and the cache hit is 100%, but it is not an option for us, we cannot rely on this. I will open an issue pinging the Windows community.

**Reply by jrosdahl** [reply]
> I actually set CCACHE_NOHASHDIR and the cache hit is 100%, but it is not an option for us, we cannot rely on this. 

Not sure I understand what you mean here. Do you mean that you don't want to risk getting object files with an incorrect embedded path to the source file? If so, didn't you have the same problem when using clcache?

**Reply by jrosdahl** [reply]
Also, did you try my tip about [enabling CCACHE_DEBUG](https://ccache.dev/manual/4.6.1.html#_cache_debugging)? That's what needs to be done to get the answer why you get cache misses.

Alternatively, you can create a minimal way for somebody else to reproduce the problem. It's hard to replicate the situation just from the non-concrete information you wrote in _Context_.

**Comment by Esysc** [comment]
> Not sure I understand what you mean here. Do you mean that you don't want to risk getting object files with an incorrect embedded path to the source file? If so, didn't you have the same problem when using clcache?

`clcache` seems to relativize absolute path correctly, I'll investigate this point:

<details>
<summary>clcache statistics:</summary>

```
"Compiler cache statistics"
  current cache dir         : Disk cache at C:\Users\concourse\worker\containers\00002e4eqsb\tmp\build\734c03a2\clcache
  cache size                : 292,130,889 bytes
  maximum cache size        : 5,122,903,040 bytes
  cache entries             : 882
  cache hits                : 878
  cache misses
    total                      : 2
    evicted                    : 0
    header changed             : 0
    source changed             : 2
  passed to real compiler
    called w/ invalid argument : 0
    called for preprocessing   : 0
    called for linking         : 0
    called for external debug  : 0
    called w/o source          : 0
    called w/ multiple sources : 0
    called w/ PCH              : 0
 ```
</details>


> Also, did you try my tip about [enabling CCACHE_DEBUG](https://ccache.dev/manual/4.6.1.html#_cache_debugging)? That's what needs to be done to get the answer why you get cache misses.

Yes sure, every run is in debug mode:

<details>
<summary>ccache configuration</summary>

```
[2022-05-24T14:51:32.833915 4940 ] === CCACHE 4.6.1 STARTED =========================================
[2022-05-24T14:51:32.833915 4940 ] Config: (default) absolute_paths_in_stderr = false
[2022-05-24T14:51:32.833915 4940 ] Config: (environment) base_dir = C:/Users/concourse/worker/containers/00001p4nuag/tmp/build/734c03a2
[2022-05-24T14:51:32.833915 4940 ] Config: (environment) cache_dir = C:\Users\concourse\worker\containers\00001p4nuag\tmp\build\734c03a2\ccache
[2022-05-24T14:51:32.833915 4940 ] Config: (default) compiler = 
[2022-05-24T14:51:32.833915 4940 ] Config: (default) compiler_check = mtime
[2022-05-24T14:51:32.833915 4940 ] Config: (default) compiler_type = auto
[2022-05-24T14:51:32.833915 4940 ] Config: (default) compression = true
[2022-05-24T14:51:32.833915 4940 ] Config: (default) compression_level = 0
[2022-05-24T14:51:32.833915 4940 ] Config: (default) cpp_extension = 
[2022-05-24T14:51:32.833915 4940 ] Config: (environment) debug = true
[2022-05-24T14:51:32.833915 4940 ] Config: (environment) debug_dir = C:\Users\concourse\worker\containers\00001p4nuag\tmp\build\734c03a2\ccachedebug
[2022-05-24T14:51:32.833915 4940 ] Config: (default) depend_mode = false
[2022-05-24T14:51:32.833915 4940 ] Config: (default) direct_mode = true
[2022-05-24T14:51:32.833915 4940 ] Config: (default) disable = false
[2022-05-24T14:51:32.833915 4940 ] Config: (default) extra_files_to_hash = 
[2022-05-24T14:51:32.833915 4940 ] Config: (default) file_clone = false
[2022-05-24T14:51:32.833915 4940 ] Config: (default) hard_link = false
[2022-05-24T14:51:32.833915 4940 ] Config: (default) hash_dir = true
[2022-05-24T14:51:32.833915 4940 ] Config: (default) ignore_headers_in_manifest = 
[2022-05-24T14:51:32.833915 4940 ] Config: (default) ignore_options = 
[2022-05-24T14:51:32.833915 4940 ] Config: (default) inode_cache = false
[2022-05-24T14:51:32.833915 4940 ] Config: (default) keep_comments_cpp = false
[2022-05-24T14:51:32.833915 4940 ] Config: (default) limit_multiple = 0.8
[2022-05-24T14:51:32.833915 4940 ] Config: (default) log_file = 
[2022-05-24T14:51:32.833915 4940 ] Config: (default) max_files = 0
[2022-05-24T14:51:32.833915 4940 ] Config: (default) max_size = 5.0G
[2022-05-24T14:51:32.833915 4940 ] Config: (default) namespace = 
[2022-05-24T14:51:32.833915 4940 ] Config: (default) path = 
[2022-05-24T14:51:32.833915 4940 ] Config: (default) pch_external_checksum = false
[2022-05-24T14:51:32.833915 4940 ] Config: (default) prefix_command = 
[2022-05-24T14:51:32.833915 4940 ] Config: (default) prefix_command_cpp = 
[2022-05-24T14:51:32.833915 4940 ] Config: (default) read_only = false
[2022-05-24T14:51:32.833915 4940 ] Config: (default) read_only_direct = false
[2022-05-24T14:51:32.833915 4940 ] Config: (default) recache = false
[2022-05-24T14:51:32.833915 4940 ] Config: (default) reshare = false
[2022-05-24T14:51:32.833915 4940 ] Config: (default) run_second_cpp = true
[2022-05-24T14:51:32.833915 4940 ] Config: (default) secondary_storage = 
[2022-05-24T14:51:32.833915 4940 ] Config: (default) sloppiness = 
[2022-05-24T14:51:32.833915 4940 ] Config: (default) stats = true
[2022-05-24T14:51:32.833915 4940 ] Config: (default) stats_log = 
[2022-05-24T14:51:32.833915 4940 ] Config: (default) temporary_dir = C:\Users\concourse\worker\containers\00001p4nuag\tmp\build\734c03a2\ccache/tmp
[2022-05-24T14:51:32.833915 4940 ] Config: (default) umask =
```
</details>

> Alternatively, you can create a minimal way for somebody else to reproduce the problem. It's hard to replicate the situation just from the non-concrete information you wrote in Context.

I'll create a simple example to easy reproduce.

**Reply by jrosdahl** [reply]
> `clcache` seems to relativize absolute path correctly, I'll investigate this point:

OK.

> > Also, did you try my tip about [enabling CCACHE_DEBUG](https://ccache.dev/manual/4.6.1.html#_cache_debugging)? That's what needs to be done to get the answer why you get cache misses.
> 
> Yes sure, every run is in debug mode:

Right, but it's not the log files that are of interest but the `.ccache-input-text` files.

> I'll create a simple example to easy reproduce.

That would great.

**Comment by Esysc** [comment]
### How to reproduce
1. Create 2 distinct working directory. The directory tree looks like below in my VirtualBox test VM:
```
C:\Users\vagrant\Desktop\test\
├───build1
│   └───tmp
│       └───build
│           ├───ccache
│           ├───debug
│           ├───includes
│           └───src
└───build2
    └───tmp
        └───build
            ├───ccache
            ├───debug
            ├───includes
            └───src
```
2. Put the same code: `HelloUniverse.cpp` in `src` and an empty header file `HelloUniverse.h` in `includes` .
```
> type HelloUniverse.cpp
#include <iostream>
#include <HelloUniverse.h>
int main(){
        std::cout << "Hello Universe" << std::endl;
        return 0;
}
```

3. cd into the first working directory and set `ccache` environment variables:
```
cd C:\Users\vagrant\Desktop\test\build1\tmp\build
set CCACHE_DIR=%cd%\ccache
set CCACHE_BASEDIR=%cd%
set CCACHE_DEBUG=true
set CCACHE_DEBUGDIR=%cd%\debug
```
4. Check `ccache` settings (I omitted default values):
```
(environment) base_dir = C:/Users/vagrant/Desktop/test/build1/tmp/build
(environment) cache_dir = C:\Users\vagrant\Desktop\test\build1\tmp\build\ccache
(environment) debug = true
(environment) debug_dir = C:\Users\vagrant\Desktop\test\build1\tmp\build\debug
```
5. run `ccache -z`
6. Compile the code
```
ccache cl.exe -I"C:\Program Files (x86)\Windows Kits\10\Include\10.0.19041.0\ucrt" -I"C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.29.30133\include" -IC:\Users\vagrant\Desktop\test\build1\tmp\build\includes /EHsc -c C:\Users\vagrant\Desktop\test\build1\tmp\build\src\HelloUniverse.cpp
```
7. Copy the content of build1 `ccache`directory into `C:\Users\vagrant\Desktop\test\build2\tmp\build\ccache>`
8. cd into the second working directory and set `ccache` environment variables:
```
cd C:\Users\vagrant\Desktop\test\build2\tmp\build
set CCACHE_DIR=%cd%\ccache
set CCACHE_BASEDIR=%cd%
set CCACHE_DEBUG=true
set CCACHE_DEBUGDIR=%cd%\debug
```
9. Compile again the code
```
ccache cl.exe -I"C:\Program Files (x86)\Windows Kits\10\Include\10.0.19041.0\ucrt" -I"C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.29.30133\include" -IC:\Users\vagrant\Desktop\test\build2\tmp\build\includes /EHsc -c C:\Users\vagrant\Desktop\test\build2\tmp\build\src\HelloUniverse.cpp
```
10. Results 100% cache is missed. 


If the same build run twice in the same `path`, cache hits 100%.
Attached  `ccache-log` and `input-text` files renamed for GitHub compliance and to identify the builds.
[build1-ccache.log](https://github.com/ccache/ccache/files/8908877/build1-ccache.log)
[build1-ccache-input-text.txt](https://github.com/ccache/ccache/files/8908892/build1-ccache-input-text.txt)
[build2-ccache.log](https://github.com/ccache/ccache/files/8908878/build2-ccache.log)
[build2-ccache-input-text.txt](https://github.com/ccache/ccache/files/8908887/build2-ccache-input-text.txt)

**Comment by vvainola** [comment]
The basedir is set within the build directory but it should be a parent folder that contains both build folders. In build1-ccache.log `Config: (environment) base_dir = C:/Users/vagrant/Desktop/test/build1/tmp/build` and in build2-ccache.log `Config: (environment) base_dir = C:/Users/vagrant/Desktop/test/build2/tmp/build`.

Without CCACHE_NOHASHDIR the ccache-input-text always contains `### source path` line which will also cause build directories to be unshareable. I think CCACHE_NOHASHDIR must be used with basedir on windows.

**Reply by Esysc** [reply]
> The basedir is set within the build directory but it should be a parent folder that contains both build folders. 

I changed the basedir to the parent of build1 and build2, the result is the same, `### source path` does the difference.
Thank you @vvainola for your comment.

**Reply by jrosdahl** [reply]
> I changed the basedir to the parent of build1 and build2, the result is the same,

That's expected. You don't _have_ to use a parent of the build/source directories – it's just that it may be convenient to do so since you then can use the same value regardless of build directory names, etc. In your case it could for instance be set to `C:\Users`.

> `### source path` does the difference.

Right, what I referred to with [_"You can opt out of this by setting CCACHE_NOHASHDIR"_](https://github.com/ccache/ccache/discussions/1084#discussioncomment-2896040). But you said that you couldn't rely on this.

Question: If you have a simple (e.g. empty) source file called test.c and run something like `clcache.exe cl.exe /c test.c` in one build directory and then the same command in another build directory, then you get a cache hit for the second compilation, right? If so, do the first and second object file contain exactly the same bytes?

**Comment by Esysc** [comment]
Sorry for late response.

> Question: If you have a simple (e.g. empty) source file called test.c and run something like clcache.exe cl.exe /c test.c in one build directory and then the same command in another build directory, then you get a cache hit for the second compilation, right? If so, do the first and second object file contain exactly the same bytes?

I followed your suggestion using 2 different paths:
1. `C:\Users\Vagrant\Desktop\build`
2. `C:\Users\Vagrant\Desktop\tests\build`

Results:

*  `CCACHE_NOHASHDIR` true: got a cache hit and the files are identical, 513 bytes same checksum.
*  `CCACHE_NOHASHDIR` false: no cache hit and files differs. The `obj` file in path 1, is 509 bytes, 4 less bytes.

**Reply by jrosdahl** [reply]
(It would be great if you could reply to my previous reply instead of creating a new comment each time. The discussion is now spread over 6 different comments instead of 1.... :slightly_smiling_face:)

OK. However, what I wanted to know was what happens when you use **clcache**, not ccache.

That is, if you do something like

```
mkdir a
cd a
echo >test.c
clcache.exe cl.exe /c test.c

cd ..
mkdir b
cd b
echo >test.c
clcache.exe cl.exe /c test.c
```

will the second call be a cache hit and will `a\test.obj` and `b\test.obj` be identical?

**Reply by Esysc** [reply]
> (It would be great if you could reply to my previous reply instead of creating a new comment each time. The discussion is now spread over 6 different comments instead of 1.... slightly_smiling_face)

Right, sorry for this scattered discussion.

> OK. However, what I wanted to know was what happens when you use **clcache**, not ccache.

Multitasking error :slightly_smiling_face: 

> 
> That is, if you do something like
> 
> ```
> mkdir a
> cd a
> echo >test.c
> clcache.exe cl.exe /c test.c
> 
> cd ..
> mkdir b
> cd b
> echo >test.c
> clcache.exe cl.exe /c test.c
> ```
> 
> will the second call be a cache hit and will `a\test.obj` and `b\test.obj` be identical?

Yes it is, I confirm the objects are identical and I've got a cache hit. 

Just for everyone to reproduce (at least in my environment):
```
mkdir a
mkdir b
fsutil file createnew a\test.c 0
fsutil file createnew b\test.c 0
call vcvars64.bat
cd a
clcache /c test.c

cd ..\b
clcache /c test.c
```

**Reply by jrosdahl** [reply]
> Yes it is, I confirm the objects are identical and I've got a cache hit.

OK, thanks! This means that clcache seems to behave exactly the same as ccache does with `CCACHE_NOHASHDIR`: both produce cache hits and neither reproduce the difference that should be between `a\test.obj` and `b\test.obj` (the embedded directory name). (In ccache's case it's intentional because of opting out with `CCACHE_NOHASHDIR`.)

What I don't understand is why you think that this is OK with clcache but that ccache with `CCACHE_NOHASHDIR` _"is not an option for us, we cannot rely on this"_.

**Reply by Esysc** [reply]
> What I don't understand is why you think that this is OK with clcache but that ccache with `CCACHE_NOHASHDIR` _"is not an option for us, we cannot rely on this"_.

This is a point that I'm going to investigate. Moreover, I want to drive some build tests on all the platforms we currently support to analyze if the behavior is the same. On Linux for example we get cache hits, but builds run always in the same path on different hosts.
I will report here.

**Comment by Esysc** [comment]
Thanks @jrosdahl and @vvainola for your time and yours exhaustive answers. After all this time, finally we decided to get rid of `clcache` and to use `ccache`. Now all supported platforms are aligned and on Windows works like a charm. I'd want just to notify you, thanks again.
